### ISSUE TITLE
Certain lat/lngs, zoom and styles throws error: Cannot read properties of null (reading 'point')

### ISSUE THREAD

**Issue by willheslam** [maplibre/maplibre-gl-js#1224]
**maplibre-gl-js version**: 2.1.9

**browser**: Chrome 100.0.4896.127 on OSX

### Steps to Trigger Behavior

Certain lng lat coordinates, with certain zoom settings, e.g.
```
{
        style: 'https://api.maptiler.com/maps/basic/style.json?key=get_your_own_OpIi9ZULNHzrESv6T2vL'
        center: [4.908376, 52.366014],
        zoom: 20
}
```
cause maplibre to throw an error whilst rendering:
```
Uncaught TypeError: Cannot read properties of null (reading 'point')
```

It seems to be dependent on specific latlngs and zoom and styles.
For example, zoom 19 or 21 doesn't trigger the error, nor does 19.9 or 20.6 (but 20.5 does)
Using the 'streets' style 
`'https://api.maptiler.com/maps/streets/style.json?key=get_your_own_OpIi9ZULNHzrESv6T2vL'`
doesn't trigger the error but 'basic' does.

Here's a screenshot from jsfiddle:
 
<img width="1792" alt="Screenshot 2022-05-04 at 09 40 33" src="https://user-images.githubusercontent.com/5377213/166649175-bba11651-9b03-4754-8d68-5748ce1345a9.png">

and here's a screenshot from the app I first encounted this bug with (using react-map-gl as a layer on top of maplibre-gl)

<img width="972" alt="Screenshot 2022-05-04 at 09 43 03" src="https://user-images.githubusercontent.com/5377213/166649285-e17554b6-853e-4614-b1e1-12c452f52b93.png">

It's specifically coming from these lines:
```
    for (const glyph of placedGlyphs) {
        addDynamicAttributes(dynamicLayoutVertexArray, glyph.point, glyph.angle);
    }
```
Logging out `placedGlyphs` results in
```
[
    {
        "point": {
            "x": 12304.450428946711,
            "y": 23700.84536864106
        },
        "angle": 8.59940111024864,
        "path": [
            {
                "x": 12320,
                "y": 23684
            },
            {
                "x": 12304.450428946711,
                "y": 23700.84536864106
            }
        ]
    },
   ...
    null,
    null,
    null,
    null,
   ...
    {
        "point": {
            "x": 12352.048734854137,
            "y": 23649.28053724135
        },
        "angle": 2.316215803069055,
        "path": [
            {
                "x": 12320,
                "y": 23684
            },
            {
                "x": 12352.048734854137,
                "y": 23649.28053724135
            }
        ]
    }
]
```
and the `null` items come specifically from this line:
```
            placedGlyphs.push(placeGlyphAlongLine(fontScale * glyphOffsetArray.getoffsetX(glyphIndex), lineOffsetX, lineOffsetY, flip, anchorPoint, tileAnchorPoint, symbol.segment,
                lineStartIndex, lineEndIndex, lineVertexArray, labelPlaneMatrix, projectionCache, rotateToLine));
 ```

inside the function `placeGlyphAlongLine` this condition is occurring:
```
        // offset does not fit on the projected line
        if (currentIndex < lineStartIndex || currentIndex >= lineEndIndex)
            return null;
```

It's unclear if this is ever expected to happen as I don't have a good understanding of the code, but given items inside `placedGlyphs` can in some circumstances be null, `addDynamicAttributes(dynamicLayoutVertexArray, glyph.point, glyph.angle);` is unsafe - either there shouldn't be nulls in the array, or nulls need to be checked for and handled appropriately - I suspect the former though I'm not sure what the context is.

My best guess is there is some kind of geometry at the coords and zoom level I described that triggers this problem.

 
### Link to Demonstration

This JS bin demonstrates the problem, but you'll need to open the dev console to see the error:

https://jsbin.com/jociwic/1/edit?html,output

You should see:
<img width="1133" alt="Screenshot 2022-05-04 at 12 41 31" src="https://user-images.githubusercontent.com/5377213/166674556-5dbc12bd-133d-4d83-b9ee-7a249f4dcc12.png">

or alternatively with jsbin:

 Create a JS fiddle from https://maplibre.org/maplibre-gl-js-docs/example/simple-map/
 and set the JS to
 ```
    var map = new maplibregl.Map({
        container: 'map',
        style: 'https://api.maptiler.com/maps/basic/style.json?key=get_your_own_OpIi9ZULNHzrESv6T2vL',
        center: [4.908376, 52.366014],
        zoom: 20   			
    });
 ```

**Comment by HarelM** [comment]
placeGlyphALongALine was recently changed in order to allow placing symbols up right for every letter.
Can you check if you can reproduce this in earlier versions?

**Comment by willheslam** [comment]
With version s2.1.8, 2.1.7, 2.1.6, 2.1.1, 2.1.0, 2.0.0, 1.15.2, 1.14.0, I get the same error

**Comment by HarelM** [comment]
Ok, thanks, this means this is not new.
If you have more time to investigate why point is null it would help a lot to push this forward.

**Comment by HarelM** [comment]
@klokan can you check if maptiler API key is blocking jsbin? I can't seem to be able to see the provided example.
@willheslam let me know if you can reproduce this with a simpler style, maybe a geojson that demonstrates this exact problem...?

**Comment by ragnarok56** [comment]
We ran into this same problem today, so this is still an issue with 4.7.1.  Troubleshooting ended up with the same conclusion as @willheslam, its unsafe to assume that first/last glyphs being placed is sufficient to render the remaining glyphs. 

In our particular case, we're using openmaptiles styles and I was able to update the style for the name attributes for the layer that was causing the error from `{name:latin}\n{name:nonlatin}` to just `{name:latin}`  (for reference, https://github.com/openmaptiles/osm-bright-gl-style/blob/master/style.json#L1871) and that seemed to fix it, but its a feels bad type of fix because I'm not sure where else this could pop up unexpectedly.

**Comment by mourner** [comment]
Just wanted to chime in and say it would be very nice to land @ragnarok56's fix — I can reproduce the issue very easily with a basic map, zooming in and out for a few seconds:

```js
const map = new maplibregl.Map({
    style: 'https://tiles.openfreemap.org/styles/dark',
    center: [30.5, 50.5],
    zoom: 7,
    container: 'map',
  })
```

**Comment by HarelM** [comment]
Definitely, but this PR requires a test to be added in order to make sure this issue doesn't return in the future...

**Comment by mourner** [comment]
@HarelM it's difficult to add a test in this case — in the [equivalent fix we did in Mapbox GL JS](https://github.com/mapbox/mapbox-gl-js/pull/12376), we didn't either; however, we hardened typing around that code, which is enough to prevent similar issues in the future. Currently, `placeGlyphAlongLine` can return a `null`, but you can make sure `placedGlyphs` array doesn't contain `null` values by typing it more strictly.

**Comment by HarelM** [comment]
Harden typing is a good approach, but adding even a unit test that passes through this new code is a better one in my opinion. 
Not adding any tests will make this code path vulnerable for future changes, stronger typings will make sure no is not returned but will not guard for future changes and will not help a person changing this part of the code create a test that passes through it, which is also not great.

**Comment by mourner** [comment]
@HarelM I get your concern, although in my experience, there are times where high standards need to be flexible, otherwise we can end up with a situation where a significant bug with a known fix doesn't get addressed for half a year and who knows for how much longer...

**Comment by HarelM** [comment]
If this is a significant bug fix I would expect someone from the community to pick it up and push this last mile forward. Having said that, I do try to help pushing PRs forward and it shouldn't been done by Isaac, but, unfortunately, it didn't happen. 
Thanks for the reminder, nevertheless, I'll look into it.

**Comment by HarelM** [comment]
BTW, here's a PR with a failing test:
- https://github.com/maplibre/maplibre-gl-js/pull/5866

**Comment by HarelM** [comment]
I've merged my branch into the fork. The fix should be available in the next version that will be released.

**Comment by mourner** [comment]
@HarelM thank you! 🙏

**Comment by HarelM** [comment]
Sure thing, feel free to copy the test if need into mapbox to keep this changes covered, I used the example you linked and added a `debugger` call into the if to know exactly when this happens, and when it did I got the map zoom and center for the test, which made it reproducible every time, so I was able to remove all the non relevant layers and sources and keep only what's needed to reproduce this issue, which was easier than trying to mock the input for the method, as there are no other unit tests for this class, which is not great, but at least I was able to add a test.
